### PR TITLE
[SPARK-22361][SQL][TEST] Add unit test for Window Frames

### PR DESCRIPTION
**Pull request by gaborgsomogyi**
## What changes were proposed in this pull request?

There are already quite a few integration tests using window frames, but the unit tests coverage is not ideal.

In this PR the already existing tests are reorganized, extended and where gaps found additional cases added.

## How was this patch tested?

Automated: Pass the Jenkins.

### CODE REVIEW

**Comment by gaborgsomogyi** [comment]
cc @jiangxb1987 @gatorsmile @hvanhovell

**Comment by gatorsmile** [comment]
ok to test

**Review by smurakozi — CHANGES_REQUESTED** [review]
Add a test checking that rangeBetween throws an AnalysisException if it gets a non-numeric column as input, as demonstrated by 
` Seq("a").toDF("k").select(min('k).over(Window.orderBy('k).rangeBetween(-1, 1))).show `

Add a test checking that rangeBetween throws an AnalysisException if it gets a column ordered by multiple order by expressions, as demonstrated by 
` Seq((1, 2)).toDF("k1", "k2").select(min('k1).over(Window.orderBy('k1, 'k2).rangeBetween(-1, 1))).show`

Add tests checking that the above mentioned rules are ignored if range is not bounded, as demonstrated by 
`Seq((1, 2)).toDF("k1", "k2").select(min('k1).over(Window.orderBy('k1, 'k2).rangeBetween(Window.unboundedPreceding, Window.unboundedFollowing))).show`

and 
`Seq("a").toDF("k").select(min('k).over(Window.orderBy('k).rangeBetween(Window.unboundedPreceding, Window.unboundedFollowing))).show`

**Comment by gaborgsomogyi** [comment]
@smurakozi nice catch, added them. Additionally found a nit.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameWindowFramesSuite.scala [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+
+    val e = intercept[AnalysisException](
+      df.select(
+        $"key",
+        count("key").over(
+          Window.partitionBy($"value").orderBy($"key").rowsBetween(0, 2147483648L))))
+    assert(e.message.contains("Boundary end is not a valid integer: 2147483648"))
+  }
+
+  test("range between should accept at most one ORDER BY expression when unbounded") {
```
Comment (jiangxb1987) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameWindowFramesSuite.scala]: "unbounded" -> "bounded"

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameWindowFramesSuite.scala [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+    )
+
+    val e = intercept[AnalysisException](
+      df.select(
+        min("key").over(window.rangeBetween(-1, 1))))
+    assert(e.message.contains("A range window frame with value boundaries cannot be used in a " +
+      "window specification with multiple order by expressions"))
+  }
+
+  test("range between should accept numeric values only when bounded") {
```
Comment (jiangxb1987) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameWindowFramesSuite.scala]: This is actually a TypeCoercion issue, I'm not sure whether we should cover it here or `TypeCoercionSuite`

**Review by jiangxb1987 — COMMENTED** [review]
Thanks for the effort! Only some minor issues.
It would also be great to improve the test for window frames parser, which currently is in `ExpressionParserSuite.test("window function expressions")`.

**Comment by jiangxb1987** [comment]
also cc @gatorsmile whom I believe should be interested to look at this.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameWindowFramesSuite.scala (hunk 1 above)
Comment (gaborgsomogyi) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameWindowFramesSuite.scala]: Fixed.

**Comment by gaborgsomogyi** [comment]
ExpressionParserSuite window related part is a bit reorganized and enhanced.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameWindowFramesSuite.scala (hunk 2 above)
Comment (gaborgsomogyi) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameWindowFramesSuite.scala]: I bit enhanced this area but not yet moved. What do you think, would it be good idea to move this test there? This is as window related as type coercion.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala [hunk 3]
```diff
@@ -263,21 +263,60 @@ class ExpressionParserSuite extends PlanTest {
       "sum(product + 1) over (partition by ((product / 2) + 1) order by 2)",
       WindowExpression('sum.function('product + 1),
         WindowSpecDefinition(Seq('product / 2 + 1), Seq(Literal(2).asc), UnspecifiedFrame)))
+  }
+
+  test("range/rows window function expressions") {
```
Comment (jiangxb1987) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala]: How about `window function expressions on various window frames`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala [hunk 4]
```diff
@@ -263,21 +263,60 @@ class ExpressionParserSuite extends PlanTest {
...
+        ordering: Seq[SortOrder] = Seq.empty,
+        frame: WindowFrame = UnspecifiedFrame): Expression = {
+      WindowExpression(func, WindowSpecDefinition(partitioning, ordering, frame))
+    }
 
-    // Range/Row
     val frameTypes = Seq(("rows", RowFrame), ("range", RangeFrame))
     val boundaries = Seq(
-      ("10 preceding", -Literal(10), CurrentRow),
+      ("unbounded preceding", UnboundedPreceding, CurrentRow),
```
Comment (jiangxb1987) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala]: tiny nit: could you add comments on each group of boundaries to explain what they are trying to cover?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala (hunk 3 above)
Comment (gaborgsomogyi) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala]: Just back from holiday. What do you mean exactly here? Range/rows between test with most of the combinations. What is missing?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala (hunk 4 above)
Comment (gaborgsomogyi) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala]: Change is on the way...

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala (hunk 3 above)
Comment (gaborgsomogyi) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala]: @jiangxb1987 gentle ping

Comment (jiangxb1987) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala]: sorry for the late response. Just update the test case name.

**Comment by jiangxb1987** [comment]
retest this please

**Comment by jiangxb1987** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merged to master/2.3

**Comment by gaborgsomogyi** [comment]
@gatorsmile @jiangxb1987 @smurakozi Thanks for the help!

**Pull request closed without merging**